create a blocks to mesh converter to support the development of the S…

RocksSavage opened the pull request:
…ideSetExtrusionGenerator ref #20880

<!--
If this PR is associated with an issue be sure to reference it
by including "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).

If this PR implements an enhancement that does not have an existing issue, please add the following:
-->

## Reason
<!--Why do you need this feature or what is the enhancement?-->
This generator was needed to bridge a gap in existing cabailities. The SideSetExtruderGenerator required by #refs #20880 could almost be made by stringing some other generators together, except there was no way to break apart subdomains/blocks into their own mesh. So we made this first to support that, and as a bonus if anyone needs this functionality it is there now. 


## Design
<!--A concise description (design) of the enhancement.-->
The resulting mesh has no sidesets. We are only adding features that directly support SideSetExtruderGenerator


## Impact
<!--Will the enhancement change existing APIs or add something new?-->


<!--
If this PR implements a bug fix, please create an issue for the bug and reference the issue.
-->

RocksSavage committed:
create a blocks to mesh converter generater

This was primarily created to support the development of the SideSetExtrusionGenerator

 ref #20880

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1094354/) on d860cde wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/21448/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1105293/) on c48f878 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21448/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21448/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 30ac1325379e2f48c8589406ce1d59020771e443`

moosebuild commented:
Job [Python 3.7](https://civet.inl.gov/job/1105407/) on 7efb4cd : invalidated by @RocksSavage

This run of the tests showed a failed test on a part of MOOSE not related to what this ticket does, and furthermore running the same command on my local machine showed that the test in question passed

loganharbour reviewed (changes requested):
Looks like you forgot to add `/test/tests/meshgenerators/block_to_mesh_converter_generator/mixedDimMesh.e`. Also, the images should be added to the large_media submodule instead.

RocksSavage commented:
> Also, the images should be added to the large_media submodule instead.

It looks to me like large_media is a submodule, a separate git repo. Is there anything special about how I commit it? I'm assuming that commiting it in there will... create a need for a new pull request somewhere else

moosebuild commented on large_media:
```diff
@@ -1 +1 @@
-Subproject commit e1211aaf2dec766e9a8e2da1631ee2f36ec7c5de
+Subproject commit 3d3986e2fa3dd32cd0bd17af98aad328ebfc544e
```
**Caution!** This contains a submodule update

moosebuild replied:
**Caution!** This contains a submodule update

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1107014/) on 6bc4b15 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>a8c2d0</th>
      <th colspan=4>#21448 6bc4b1</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>83.54%</td>
      <td>83.54%</td>
      <td>-0.00%</td>
      <td>92.11%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>77826</td>
      <td>77857</td>
      <td>+31</td>
      <td>35</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15334</td>
      <td>15341</td>
      <td>+7</td>
      <td>3</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/21448/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/21448/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/navier_stokes)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/richards)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/21448/coverage/xfem)

</details>



This comment will be updated on new commits.

GiudGiud committed:
Misc improvements:
- clearer docs
- reworded test requirements
- typos
Refs #20880

GiudGiud commented:
Should be ready for more review
Involving @roystgnr as this can be used to extract sidesets for Triangulation

GiudGiud commented:
Should be passing now

GiudGiud commented:
@roystgnr @loganharbour need review please

loganharbour commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+
+  auto new_mesh = buildMeshBaseObject();
+
+  std::vector<subdomain_id_type> target_block_ids =
```
```suggestion
  const auto target_block_ids =
```

loganharbour commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+    {
+      paramError("target_blocks",
+                 "The target_block '",
+                 getParam<std::vector<SubdomainName>>("target_blocks")[i],
```
```suggestion
                 _target_blocks[i],
```

loganharbour commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  // know which nodes have already been inserted, by tracking the old mesh's node's ids'
+  std::unordered_map<dof_id_type, dof_id_type> old_new_node_map;
+
+  for (subdomain_id_type target_block_id : target_block_ids)
```
```suggestion
  for (const auto target_block_id : target_block_ids)
```

loganharbour commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+      // correctly assign new copies of nodes, loop over nodes
+      for (dof_id_type i : elem->node_index_range())
+      {
+        auto & n = *elem->node_ptr(i);
```
```suggestion
        auto & n = elem->node_ref(i);
```

loganharbour commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+      new_mesh->add_elem(copy.release());
+    }
+  }
+  new_mesh->prepare_for_use();
```
Do we need to prepare?

GiudGiud replied:
I think we do for the next step which will be extrusion then stitching
that s how it came about

roystgnr replied:
Isn't this what the whole `MeshBase::is_prepared()` and `MeshBase::set_isnt_prepared()` API pair is for?  So we can avoid double-prepare calls in a sequence of generators?  A generator that needs a prepared input mesh is supposed to be testing whether that's been done already, and if not then doing it itself.

GiudGiud replied:
Makes sense.

loganharbour commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -0,0 +1,53 @@
+
+  [extract_block]
+    requirement = 'The system shall have the capability of creating a new mesh given '
+                  'one or more subdomains within an existing mesh'
```
```suggestion
                  'one or more subdomains within an existing mesh, with'
```

loganharbour commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -0,0 +1,53 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+
+      detail = 'in the simplest case - a one block, single element, 2d mesh of a square,'
```
```suggestion
      detail = 'a one block, single element, 2d mesh of a square,'
```

loganharbour commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -0,0 +1,53 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+
+      detail = 'with 2D blocks to be extracted from a 2D mesh'
```
```suggestion
      detail = '2D blocks to be extracted from a 2D mesh'
```

loganharbour commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -0,0 +1,53 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+
+      detail = 'with multiple subdomains/blocks which can be incorporated into a new mesh in 3D'
```
```suggestion
      detail = 'multiple subdomains/blocks which can be incorporated into a new mesh in 3D'
```

loganharbour commented on test/tests/meshgenerators/block_to_mesh_converter_generator/tests:
```diff
@@ -0,0 +1,53 @@
+      cli_args = '--mesh-only'
+      mesh_mode = 'replicated'
+
+      detail = 'with a lower dimension block (1D) to be extracted from a 2D mesh'
```
```suggestion
      detail = 'a lower dimension block (1D) to be extracted from a 2D mesh'
```

roystgnr commented on framework/doc/content/source/meshgenerators/BlockToMeshConverterGenerator.md:
```diff
@@ -0,0 +1,29 @@
+## Overview
+
+The `BlockToMeshConverterGenerator` is for moving one (or more) blocks from a mesh to form a new mesh object.
+This does not erase/delete the blocks from the original mesh. The new mesh has only one block (named 0). 
```
We're moving the input mesh to `mesh` below, right?  And then it gets deleted, as per usual `MeshGenerator` behavior?  But "This does not erase/delete the blocks from the original mesh." makes me imagine that the original mesh will instead somehow be preserved and available for use in a subsequent `MeshGenerator`.

I'd say delete that whole sentence, but maybe I'm just not yet understanding what was meant by it.

GiudGiud replied:
It should be available for use in another MeshGenerator. That s how we did the sideset extrusion

GiudGiud replied:
like the idea is that the mesh was not explicitly marked for deletion, and another MG will pick it up. Then whatever we did with the extracted blocks (extrusion or triangulation are planned rn) can be stitched on the initial mesh

roystgnr replied:
> It should be available for use in another MeshGenerator.

Sorry, but: can you explain how?

At `std::unique_ptr<MeshBase> mesh = std::move(_input);` the pointer to the input mesh is moved into `mesh`, which is a `unique_ptr` that still has an active deleter.  When `BlockToMeshConverterGenerator::generate()` ends, `mesh` goes out of scope, its destructor calls the deleter, and the input mesh gets deleted.  Is there something else I'm missing?

GiudGiud replied:
seems like we should avoid the move then.

roystgnr replied:
Okay, I think I found the something else I'm missing.  In `MooseApp::executeMeshGenerators()`, we're checking how many downstream generators use each output mesh, and if we see N>1 of them then we make N-1 copies so that each can have its own, and then each downstream generator is responsible for deleting its own copy.  I can see ways to make that more efficient but I no longer see any problem with correctness.  Sorry about the false alarm!

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  std::unique_ptr<MeshBase> mesh = std::move(_input);
+
+  if (!mesh->is_replicated())
+    mooseError("BlockToMeshConverterGenerator is not implemented for distributed meshes.");
```
This support would be really easy to add.  I see that this error message is a common cut-and-paste-and-edit in MOOSE (`TiledMeshGenerator` forgot to change the name in the string, even!), and indeed distributed mesh support in generators can be extremely tricky, but in this case most of the algorithm should be embarrassingly parallel and the only catch would be the need to check for and null out obsolete `remote_elem` links at the end.

Would anyone want this support?

GiudGiud replied:
the use cases for this are sideset extrusion and triangulator.
neither of these support distributed mesh currently though I could absolutely see it in the future. 

I dont want it right now tbh, I'm trying to have all my summer intern PRs merged at this time.

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+  for (subdomain_id_type target_block_id : target_block_ids)
+  {
+
+    for (auto elem : mesh->active_subdomain_elements_ptr_range(target_block_id))
```
So if we have a uniformly refined mesh, this will flatten it, and if we have an adaptively refined mesh, this will break it?  The former behavior at least would need to be clearly documented, and the latter behavior is unacceptable; at the very least we need a test that would catch that and error out.

GiudGiud replied:
I m not sure what you mean here. I just tried this
```
[Mesh]
  [file]
    type = FileMeshGenerator
    file = 3dmultiblock.e
  []
  [refine]
    type = RefineBlockGenerator
    input = file
    block = '2 4 6'
    refinement = '1 1 1'
  []
  [blockToMesh]
    type = BlockToMeshConverterGenerator
    input = refine
    target_blocks = "1 12 2 3 5 6"
  []
[]
```
and got the intended behavior afaik

roystgnr replied:
Let me try this input and see what I see.

roystgnr replied:
Okay, this time I don't think I'm mistaken; this is an actual bug.  Take a look at `n_elem()` and `n_active_elem()` in the console output.  I'm seeing 180 for both.  But in an adaptively refined mesh, those should never be the exact same number; any 2^dim level-1 active elements should have 1 non-active ancestor element, any 4^dim level-2 elements should have 2^dim+1 ancestors, etc.  Open the mesh here in paraview and it may look like an adaptively refined mesh, but libMesh will notice that those coarse elements don't have any neighbors on the side where the refined elements live, it'll happily conclude that it's looking at a slit mesh with internal boundaries at that interface, and at that point we can just pray the users don't believe their results.

GiudGiud replied:
ok good catch!

I m just going to error out on these.

We dont currently run generators during a simulation, only at the very beginning.
Refinement at the beginning could just be done after the block extraction using this new MG so this is not that limiting either.

roystgnr replied:
Yeah, just exiting with an error is fine by me.  Having to reorder mesh generators is merely annoying; a mesh that looks correct but isn't would have been a disaster.

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+          // add them both to the element and the mesh
+
+          // Nodes' IDs are their indexes in the nodes' respective mesh
+          // If we set them as invalid they are automatically re-assigned
+          // to be the last element in the mesh's node array
```
```suggestion
          // If we set them as invalid they are automatically assigned
```

Using the last index is `ReplicatedMesh` behavior but it's not guaranteed in general; `DistributedMesh` has a trickier system to enable non-conflicting parallel insertions.

roystgnr commented on framework/src/meshgenerators/BlockToMeshConverterGenerator.C:
```diff
@@ -0,0 +1,125 @@
+          // Nodes' IDs are their indexes in the nodes' respective mesh
+          // If we set them as invalid they are automatically re-assigned
+          // to be the last element in the mesh's node array
+          std::unique_ptr<Node> node_copy = Node::build(elem->point(i), Node::invalid_id);
+
+          // Add to mesh. This method call will automatically re-assign the id; so must call this
+          // first
+          new_mesh->add_node(node_copy.get());
+
+          // Add to element copy (manually)
+          copy->set_node(copy_n_index++) = node_copy.get();
+
+          // remember the (old) ID
+          old_new_node_map[n.id()] = node_copy.release()->id();
```
```suggestion
          // Add to mesh, auto-assigning a new id.
          Node * node = new_mesh->add_point(elem->point(i));

          // Add to element copy (manually)
          copy->set_node(copy_n_index++) = node;

          // remember the (old) ID
          old_new_node_map[n.id()] = node->id();
```

If you ever find yourself using `std::unique_ptr::release()`, that's a code smell.  Maybe you were really forced into it, but double-check first.

GiudGiud commented on framework/include/meshgenerators/BlockToMeshConverterGenerator.h:
```diff
@@ -0,0 +1,31 @@
+  /// Mesh that comes from another generator
+  std::unique_ptr<MeshBase> & _input;
+
+  const std::vector<SubdomainName> _target_blocks;
```
```suggestion
  /// Blocks to extract to form a mesh
  const std::vector<SubdomainName> _target_blocks;
```

GiudGiud committed:
Apply suggestions from code review
- dont prepare for use from produce MG, prepare for use from consumer MG (to come)
- cleaner pointer logic
- cleaner tests file requirements

Co-authored-by: Logan Harbour <loganhharbour@gmail.com>
Co-authored-by: roystgnr <roy@stogners.org>

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1217302/) on 82dda8a wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/21448/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/21448/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 7e9638b1a7cd2b255055a173929a9e6fbaf93da2`

GiudGiud commented:
@loganharbour your review should be addressed
@roystgnr I responded to your comments and will need a round of reply for further action

GiudGiud committed:
Error out for refined blocks as the output mesh would be disconnected and overlapped (not reproduce the initial parent-child structure)
Add exception checking for distributed meshed and refined blocks

GiudGiud commented:
Should be good now. Added two tests for these exceptions too
